### PR TITLE
Fixed 'er up

### PR DESCRIPTION
**Pull request by chrisps**

### CODE REVIEW

File: src/nc/arch/x86/X86InstructionAnalyzer.cpp [hunk 1]
```diff
@@ -134,34 +134,103 @@ class X86InstructionAnalyzerImpl {
...
+                    operand(0) ^= temporary(operand(0).size()),
                     pf ^= intrinsic(),
                     zf ^= operand(0) == constant(0),
                     sf ^= signed_(operand(0)) < constant(0),
                     of ^= intrinsic(),
                     af ^= intrinsic(),
                     less ^= ~(sf == of),
                     less_or_equal ^= less | zf,
-                    below_or_equal ^= cf | zf
+                    below_or_equal ^= cf | zf\
```
Comment (yegord) [inline src/nc/arch/x86/X86InstructionAnalyzer.cpp]: Why do you need \ ?

File: src/nc/arch/x86/X86InstructionAnalyzer.cpp [hunk 2]
```diff
@@ -134,34 +134,103 @@ class X86InstructionAnalyzerImpl {
...
-                    cf ^= intrinsic(),
-                    pf ^= intrinsic(),
-                    zf ^= operand(0) == constant(0),
-                    sf ^= signed_(operand(0)) < constant(0),
-                    of ^= intrinsic(),
-                    af ^= intrinsic(),
-                    less ^= ~(sf == of),
-                    less_or_equal ^= less | zf,
-                    below_or_equal ^= cf | zf
+                        temporary(operand(0).size()) ^= operand(0) + operand(1),
```
Comment (yegord) [inline src/nc/arch/x86/X86InstructionAnalyzer.cpp]: Why did you change indentation?

File: src/nc/arch/x86/X86InstructionAnalyzer.cpp [hunk 3]
```diff
@@ -134,34 +134,103 @@ class X86InstructionAnalyzerImpl {
...
+                        cf ^= unsigned_(temporary(operand(0).size())) < unsigned_(operand(0)),
+                        operand(0) ^= temporary(operand(0).size()),
+                        pf ^= intrinsic(),
+                        zf ^= operand(0) == constant(0),
+                        sf ^= signed_(operand(0)) < constant(0),
+                        of ^= intrinsic(),
+                        af ^= intrinsic(),
+                        less ^= ~(sf == of),
+                        less_or_equal ^= less | zf,
+                        below_or_equal ^= cf | zf\
```
Comment (yegord) [inline src/nc/arch/x86/X86InstructionAnalyzer.cpp]: Why do you need \ ?

File: src/nc/arch/x86/X86InstructionAnalyzer.cpp [hunk 4]
```diff
@@ -134,34 +134,103 @@ class X86InstructionAnalyzerImpl {
...
+            }
+            case UD_Istc: {
+                _[
+                        cf ^= constant(1),
+                                below_or_equal ^= cf | zf
+                ];
+                break;
+            }
+            case UD_Isahf: {
+#define extractFlag(offs) MemoryLocationExpression(X86Registers::ah()->memoryLocation().resized(1).shifted(offs))
```
Comment (yegord) [inline src/nc/arch/x86/X86InstructionAnalyzer.cpp]: Macros are evil. Let's make it a lambda.

File: src/nc/core/likec/Simplifier.cpp [hunk 5]
```diff
@@ -345,6 +345,22 @@ std::unique_ptr<Expression> Simplifier::simplify(std::unique_ptr<BinaryOperator>
...
     if (node->operatorKind() == BinaryOperator::ASSIGN) {
+        /*
+         * if the right hand side is a cast to the type of
+         * the left hand side, check to see if the right hand side expression
+         * will automatically be upgraded to the type of the left hand side
+         * only works for integer types.
+         */
+        if(auto src = node->right()->as<Typecast>()) {
+            auto destType = typeCalculator_.getType(node->left().get());//leftIdent->declaration()->type();
+            auto cstType = typeCalculator_.getType(src->operand().get());//src->type();
```
Comment (yegord) [inline src/nc/core/likec/Simplifier.cpp]: What is the purpose of the two comments?

File: src/nc/core/likec/TypeCalculator.h [hunk 6]
```diff
@@ -45,6 +45,9 @@ class TypeCalculator {
     const Type *getType(const VariableIdentifier *node);
     const Type *getType(const UndeclaredIdentifier *node);
     const Type *getBinaryOperatorType(int operatorKind, const Expression *left, const Expression *right);
+    Tree& getTree() {
```
Comment (yegord) [inline src/nc/core/likec/TypeCalculator.h]: We do not use 'get' prefix in getter names.

File: src/nc/core/likec/Simplifier.cpp [hunk 7]
```diff
@@ -400,7 +416,45 @@ std::unique_ptr<MemberAccessOperator> Simplifier::simplify(std::unique_ptr<Membe
 
 std::unique_ptr<Expression> Simplifier::simplify(std::unique_ptr<Typecast> node) {
     node->operand() = simplify(std::move(node->operand()));
-
+    /*
+     * reinterpret_cast<int16_t>(*reinterpret_cast<uint16_t*>(&edi2))
+     * will become *reinterpret_cast<int16_t*>(&edi2)
+     */
+    if(node->operand()->is<UnaryOperator>()
+                                          && node->type()->isInteger() &&
```
Comment (yegord) [inline src/nc/core/likec/Simplifier.cpp]: Indentation here and in many other places is completely random and inconsistent. Could you fix it?

File: src/nc/arch/x86/X86InstructionAnalyzer.cpp [hunk 8]
```diff
@@ -134,34 +134,103 @@ class X86InstructionAnalyzerImpl {
...
+                ];
+                break;
+            }
+            case UD_Ibswap: {
+                if(operand(0).size() == 32) {
+                    _[
+                            operand(0)  ^= ((unsigned_(operand(0))>>constant(24))&constant(0xFF)) |
+                            ((unsigned_(operand(0))>>constant(8))&constant(0xFF00))
+                            | ((operand(0)<<constant(8))&constant(0xFF0000))
+                            | ((operand(0)<<constant(24))&constant(0xFF000000))
```
Comment (yegord) [inline src/nc/arch/x86/X86InstructionAnalyzer.cpp]: Do you need `& 0xFF000000` here?

File: src/nc/arch/x86/X86InstructionAnalyzer.cpp [hunk 9]
```diff
@@ -134,34 +134,103 @@ class X86InstructionAnalyzerImpl {
...
+            case UD_Ifnstsw: {
+                _[
+                        operand(0) ^= regizter(X86Registers::fpu_status_word())
+                ];
+                break;
+            }
+            case UD_Ibswap: {
+                if(operand(0).size() == 32) {
+                    _[
+                            operand(0)  ^= ((unsigned_(operand(0))>>constant(24))&constant(0xFF)) |
```
Comment (yegord) [inline src/nc/arch/x86/X86InstructionAnalyzer.cpp]: Do you need `& 0xFF` here?

File: src/nc/core/likec/Simplifier.cpp [hunk 10]
```diff
@@ -400,7 +416,45 @@ std::unique_ptr<MemberAccessOperator> Simplifier::simplify(std::unique_ptr<Membe
...
+            node->operand()->as<UnaryOperator>()->operatorKind() == UnaryOperator::DEREFERENCE) {
+        /*
+         * node's operand is dereference. node is a cast to an IntegerType
```
Comment (yegord) [inline src/nc/core/likec/Simplifier.cpp]: I doubt that this comment helps anything.
The if condition already says everything in pretty good English.

File: src/nc/core/likec/Simplifier.cpp [hunk 11]
```diff
@@ -400,7 +416,45 @@ std::unique_ptr<MemberAccessOperator> Simplifier::simplify(std::unique_ptr<Membe
...
+         */
+        auto deref = node->operand()->as<UnaryOperator>();
+
+        if(deref->operand()->is<Typecast>()) {
```
Comment (yegord) [inline src/nc/core/likec/Simplifier.cpp]: `if (auto innerCast = deref->operand()->as<Typecast>())`

File: src/nc/core/likec/Simplifier.cpp [hunk 12]
```diff
@@ -400,7 +416,45 @@ std::unique_ptr<MemberAccessOperator> Simplifier::simplify(std::unique_ptr<Membe
...
+             *
+             * we're going to check to see if the type of pointer it's casted to
+             * is compatible with the type this node is casting to
+             * i.e reinterpret_cast<int16_t>(*reinterpret_cast<uint16_t*>(&edi2))
+             * will become *reinterpret_cast<int16_t*>(&edi2)
+             */
+
+            auto innerCast = deref->operand()->as<Typecast>();
+            if (innerCast->type()->isPointer() && !node->type()->isPointer()) {
+                //inner cast is a pointer type
```
Comment (yegord) [inline src/nc/core/likec/Simplifier.cpp]: `assert(ptrType);` would say the same thing, but would also allow for automatic checking at run time.

File: src/nc/core/likec/Simplifier.cpp [hunk 13]
```diff
@@ -400,7 +416,45 @@ std::unique_ptr<MemberAccessOperator> Simplifier::simplify(std::unique_ptr<Membe
...
 std::unique_ptr<Expression> Simplifier::simplify(std::unique_ptr<Typecast> node) {
     node->operand() = simplify(std::move(node->operand()));
-
+    /*
+     * reinterpret_cast<int16_t>(*reinterpret_cast<uint16_t*>(&edi2))
+     * will become *reinterpret_cast<int16_t*>(&edi2)
+     */
+    if(node->operand()->is<UnaryOperator>()
+                                          && node->type()->isInteger() &&
+            node->operand()->as<UnaryOperator>()->operatorKind() == UnaryOperator::DEREFERENCE) {
```
Comment (yegord) [inline src/nc/core/likec/Simplifier.cpp]: I would put `node->operand()` checks next to each other: related things should be closer than non-related things.

**Comment by chrisps** [comment]
Nope

**Comment by chrisps** [comment]
Nope

**Comment by chrisps** [comment]
The identation all looks very nice to me. I don't see the issue. As for the two unnecessary constants, those will be removed. The comments I forgot to remove before the commit. The get will get got, and the backslash deleted, but as it is the code works and improves the output.

**Comment by yegord** [comment]
On Wed, Mar 29, 2017 at 08:14:37PM -0700, chrisps wrote:
> The identation all looks very nice to me. I don't see the issue.

Here are some excerpts from the output of
'git diff master..chrisps/master':

Change of indentation from the correct four spaces to incorrect eight spaces:

             case UD_Iadd: {
                 _[
-                    operand(0) ^= operand(0) + operand(1),
-                    cf ^= intrinsic(),
-                    pf ^= intrinsic(),
-                    zf ^= operand(0) == constant(0),
-                    sf ^= signed_(operand(0)) < constant(0),
-                    of ^= intrinsic(),
-                    af ^= intrinsic(),
-                    less ^= ~(sf == of),
-                    less_or_equal ^= less | zf,
-                    below_or_equal ^= cf | zf
+                        temporary(operand(0).size()) ^= operand(0) + operand(1),
+                        cf ^= unsigned_(temporary(operand(0).size())) < unsigned_(operand(0)),
+                        operand(0) ^= temporary(operand(0).size()),
+                        pf ^= intrinsic(),
+                        zf ^= operand(0) == constant(0),
+                        sf ^= signed_(operand(0)) < constant(0),
+                        of ^= intrinsic(),
+                        af ^= intrinsic(),
+                        less ^= ~(sf == of),
+                        less_or_equal ^= less | zf,
+                        below_or_equal ^= cf | zf\
                 ];

Besides not looking nicely, it leads to git diff showing differences which are
actually not there, git blame showing commits which actually did not introduce
the line.

Two lines inside brackets indented by eight and sixteen spaces (stairs),
instead of all being indented by four (the same with UD_Icmc and UD_Istc).

+            case UD_Iclc: {
+                _[
+                        cf ^= constant(0),
+                                below_or_equal ^= cf | zf
+                ];
+                break;
+            }

The second line indented without any reasonable logic behind it.

+    if(node->operand()->is<UnaryOperator>()
+                                          && node->type()->isInteger() &&
+            node->operand()->as<UnaryOperator>()->operatorKind() == UnaryOperator::DEREFERENCE) {

Continuation of the statement in not indented:

+                    deref->operand() =
+                    std::make_unique<Typecast>(Typecast::CastKind::REINTERPRET_CAST,

Am I a bore? Yes, I am.

Could I fix it myself? Yes, but then you'll send me code with such problems
again, and I will have to fix it again, but I have more fun things to do.

> as it is the code works and improves the output

It's great. This is why I want it in the master and spend time reviewing it.

-- 
Yegor Derevenets

**Comment by yegord** [comment]
You can also fix some warnings:
```
[157/231] Building CXX object nc/core/CMakeFiles/nc-core.dir/likec/Simplifier.cpp.o
/home/yegor/devel/snowman/src/nc/core/likec/Simplifier.cpp:360:28: warning: moving a temporary object prevents copy elision [-Wpessimizing-move]
                    return std::move(simplify(std::move(node)));
                           ^
/home/yegor/devel/snowman/src/nc/core/likec/Simplifier.cpp:360:28: note: remove std::move call here
                    return std::move(simplify(std::move(node)));
                           ^~~~~~~~~~                         ~
/home/yegor/devel/snowman/src/nc/core/likec/Simplifier.cpp:445:26: warning: unused variable 'integerPointee' [-Wunused-variable]
                    auto integerPointee = ptrType->pointeeType()->as<IntegerType>();
```
(This is commit 7a4e52ff3c45b3cdb4a989ffa527ca1625748399 being built.)

**Comment by chrisps** [comment]
Aye aye cap'n

**Comment by yegord** [comment]
FYI, there was also
```
..\..\..\snowman\src\nc\core\likec\Simplifier.cpp(445): warning C4482: nonstandard extension used: enum 'nc::core::likec::Typecast::CastKind' used in qualified name [C:\yegor\buildslave\snowman-pr-msvc2010-32-qt-4_7\build\nc\core\nc-core.vcxproj]
  ..\..\..\snowman\src\nc\core\likec\Simplifier.cpp(445): warning C4482: nonstandard extension used: enum 'nc::core::likec::Typecast::CastKind' used in qualified name [C:\yegor\buildslave\snowman-pr-msvc2010-32-qt-4_7\build\nc\core\nc-core.vcxproj]
```
But I guess I'll fix this and the rest myself.

**Comment by chrisps** [comment]
Ah, I'm not using msvc so that may be why I did not receive that warning.

**Comment by yegord** [comment]
Merged. Thanks!
